File: freightmas/forwarding_service/report/forwarding_job_closing_report/forwarding_job_closing_report.py
# Copyright (c) 2026, Zvomaita Technologies (Pvt) Ltd and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import formatdate


def execute(filters=None):
    if not filters:
        filters = {}

    columns = get_columns()
    data = get_data(filters)

    return columns, data


def get_columns():
    return [
        {
            "label": _("Job ID"),
            "fieldname": "job_id",
            "fieldtype": "Link",
            "options": "Forwarding Job",
            "width": 150,
        },
        {
            "label": _("Customer"),
            "fieldname": "customer",
            "fieldtype": "Link",
            "options": "Customer",
            "width": 200,
        },
        {
            "label": _("Customer Reference"),
            "fieldname": "customer_reference",
            "fieldtype": "Data",
            "width": 150,
        },
        {
            "label": _("Completed On"),
            "fieldname": "completed_on",
            "fieldtype": "Data",
            "width": 110,
        },
        {
            "label": _("Revenue Recognised"),
            "fieldname": "revenue_recognised",
            "fieldtype": "Data",
            "width": 130,
        },
        {
            "label": _("RR Date"),
            "fieldname": "revenue_recognised_on",
            "fieldtype": "Data",
            "width": 110,
        },
        {
            "label": _("Status"),
            "fieldname": "status",
            "fieldtype": "Data",
            "width": 110,
        },
    ]


def get_data(filters):
    conditions = [
        "fj.status IN ('Completed', 'Closed')",
        "fj.docstatus < 2",
    ]
    params = {}

    if filters.get("status"):
        conditions.append("fj.status = %(status)s")
        params["status"] = filters["status"]

    if filters.get("customer"):
        conditions.append("fj.customer = %(customer)s")
        params["customer"] = filters["customer"]

    where_clause = " AND ".join(conditions)

    jobs = frappe.db.sql(
        """
        SELECT
            fj.name,
            fj.customer,
            fj.customer_reference,
            fj.completed_on,
            fj.revenue_recognised,
            fj.revenue_recognised_on,
            fj.status
        FROM `tabForwarding Job` fj
        WHERE {where_clause}
        ORDER BY fj.completed_on DESC
    """.format(
            where_clause=where_clause
        ),
        params,
        as_dict=True,
    )

    data = []
    for job in jobs:
        data.append(
            {
                "job_id": job.name,
                "customer": job.customer or "",
                "customer_reference": job.get("customer_reference") or "",
                "completed_on": format_date(job.get("completed_on")),
                "revenue_recognised": "Yes" if job.get("revenue_recognised") else "No",
                "revenue_recognised_on": format_date(job.get("revenue_recognised_on")),
                "status": job.get("status") or "",
            }
        )

    return data


def format_date(date_value):
    if not date_value:
        return ""
    try:
        return formatdate(date_value, "dd-MMM-yy")
    except Exception:
        return str(date_value)

File: freightmas/forwarding_service/report/forwarding_job_closing_report/forwarding_job_closing_report.js
// Copyright (c) 2026, Zvomaita Technologies (Pvt) Ltd and contributors
// For license information, please see license.txt

frappe.query_reports["Forwarding Job Closing Report"] = {
	"filters": [
		{
			fieldname: "status",
			label: __("Status"),
			fieldtype: "Select",
			options: "\nCompleted\nClosed",
			default: ""
		},
		{
			fieldname: "customer",
			label: __("Customer"),
			fieldtype: "Link",
			options: "Customer"
		}
	],

	onload: function(report) {
		// Standard Excel Export
		report.page.add_inner_button('Export to Excel', function() {
			const filters = report.get_filter_values(true);
			const query = encodeURIComponent(JSON.stringify(filters));
			const url = `/api/method/freightmas.api.export_report_to_excel?report_name=Forwarding Job Closing Report&filters=${query}`;
			window.open(url);
		}, 'Export');

		// Clear Filters
		report.page.add_inner_button('Clear Filters', function() {
			report.filters.forEach(filter => {
				let default_value = filter.df.default || "";
				report.set_filter_value(filter.df.fieldname, default_value);
			});
			report.refresh();
		});
	},

	formatter: function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		// Color-code status
		if (column.fieldname === "status" && data) {
			let color = "gray";
			if (data.status === "Completed") color = "orange";
			else if (data.status === "Closed") color = "green";

			value = `<span style="color: ${color}; font-weight: bold;">${value}</span>`;
		}

		// Color-code revenue recognised
		if (column.fieldname === "revenue_recognised" && data) {
			let color = data.revenue_recognised === "Yes" ? "green" : "red";
			value = `<span style="color: ${color}; font-weight: bold;">${value}</span>`;
		}

		return value;
	}
};
